File: src/lib/commands/prReview.ts
import { z } from "zod";
import {
  invokeCommand,
  mentionCandidatesSchema,
  MentionCandidate,
  prFileDiffSchema,
  PrFileDiff,
} from "./runtime";

const prReviewerSchema = z.object({
  id: z.string().nullable(),
  displayName: z.string(),
  vote: z.number(),
  voteLabel: z.string(),
  isRequired: z.boolean(),
  isMe: z.boolean(),
});

export type PrReviewer = z.infer<typeof prReviewerSchema>;

const prCommentSchema = z.object({
  id: z.number(),
  parentCommentId: z.number().nullable(),
  content: z.string().nullable(),
  author: z.string().nullable(),
  publishedDate: z.string().nullable(),
  isSystem: z.boolean(),
  isMine: z.boolean(),
});

export type PrComment = z.infer<typeof prCommentSchema>;

const prThreadSchema = z.object({
  id: z.number(),
  status: z.string().nullable(),
  isResolved: z.boolean(),
  filePath: z.string().nullable(),
  rightLine: z.number().nullable(),
  leftLine: z.number().nullable().default(null),
  comments: z.array(prCommentSchema),
});

export type PrThread = z.infer<typeof prThreadSchema>;

const pullRequestReviewSchema = z.object({
  pullRequestId: z.number(),
  title: z.string(),
  description: z.string().nullable(),
  sourceRefName: z.string(),
  targetRefName: z.string(),
  createdBy: z.string().nullable(),
  creationDate: z.string().nullable(),
  isDraft: z.boolean(),
  autoComplete: z.boolean().default(false),
  reviewers: z.array(prReviewerSchema),
  threads: z.array(prThreadSchema),
});

export type PullRequestReview = z.infer<typeof pullRequestReviewSchema>;

const prCommitSchema = z.object({
  commitId: z.string(),
  shortCommitId: z.string(),
  comment: z.string(),
  authorName: z.string().nullable(),
  authorDate: z.string().nullable(),
  webUrl: z.string().nullable(),
});

const prCommitsSchema = z.array(prCommitSchema);

export type PrCommit = z.infer<typeof prCommitSchema>;

const prChangedFileSchema = z.object({
  path: z.string(),
  changeType: z.string(),
  originalPath: z.string().nullable(),
});

export type PrChangedFile = z.infer<typeof prChangedFileSchema>;

const pullRequestChangesSchema = z.object({
  baseCommitId: z.string().nullable(),
  targetCommitId: z.string().nullable(),
  files: z.array(prChangedFileSchema),
});

export type PullRequestChanges = z.infer<typeof pullRequestChangesSchema>;

const prStatusResultSchema = z.object({
  status: z.string().nullable(),
  isDraft: z.boolean(),
});
export type PrStatusResult = z.infer<typeof prStatusResultSchema>;

export type PullRequestAction =
  | "abandon"
  | "reactivate"
  | "publish"
  | "complete"
  | "enableAutoComplete"
  | "cancelAutoComplete";

const prDetailsResultSchema = z.object({
  title: z.string(),
  description: z.string().nullable(),
});
export type PrDetailsResult = z.infer<typeof prDetailsResultSchema>;

export type PrLocatorInput = {
  organizationId?: string;
  projectId: string;
  repositoryId: string;
  pullRequestId: number;
};

/** Builds the PR locator shared by every PR review command. */
export function prLocator(pr: {
  organizationId: string;
  projectId: string;
  repositoryId: string;
  pullRequestId: number;
}): PrLocatorInput {
  return {
    organizationId: pr.organizationId,
    projectId: pr.projectId,
    repositoryId: pr.repositoryId,
    pullRequestId: pr.pullRequestId,
  };
}

export type GetPullRequestReviewInput = PrLocatorInput;
export type ListPullRequestChangesInput = PrLocatorInput;
export type ListPullRequestCommitsInput = PrLocatorInput;

export type GetPullRequestFileDiffInput = PrLocatorInput & {
  filePath: string;
  originalPath?: string | null;
  changeType: string;
  baseCommitId?: string | null;
  targetCommitId?: string | null;
};

export type PostPullRequestCommentInput = PrLocatorInput & {
  threadId?: number;
  content: string;
  filePath?: string;
  rightLine?: number;
  leftLine?: number;
};

export type SetPullRequestThreadStatusInput = PrLocatorInput & {
  threadId: number;
  status: "active" | "closed";
};

export type SubmitPullRequestVoteInput = PrLocatorInput & {
  vote: -10 | -5 | 0 | 5 | 10;
};

export type SearchPullRequestMentionsInput = {
  organizationId?: string;
  query: string;
};

export type EditPullRequestCommentInput = PrLocatorInput & {
  threadId: number;
  commentId: number;
  content: string;
};

export type DeletePullRequestCommentInput = PrLocatorInput & {
  threadId: number;
  commentId: number;
};

export async function getPullRequestReview(
  input: GetPullRequestReviewInput,
): Promise<PullRequestReview> {
  const result = await invokeCommand("get_pull_request_review", { input });
  return pullRequestReviewSchema.parse(result);
}

export async function listPullRequestChanges(
  input: ListPullRequestChangesInput,
): Promise<PullRequestChanges> {
  const result = await invokeCommand("list_pull_request_changes", { input });
  return pullRequestChangesSchema.parse(result);
}

export async function listPullRequestCommits(
  input: ListPullRequestCommitsInput,
): Promise<PrCommit[]> {
  const result = await invokeCommand("list_pull_request_commits", { input });
  return prCommitsSchema.parse(result);
}

export async function getPullRequestFileDiff(
  input: GetPullRequestFileDiffInput,
): Promise<PrFileDiff> {
  const result = await invokeCommand("get_pull_request_file_diff", { input });
  return prFileDiffSchema.parse(result);
}

export async function postPullRequestComment(
  input: PostPullRequestCommentInput,
): Promise<PrThread> {
  const result = await invokeCommand("post_pull_request_comment", { input });
  return prThreadSchema.parse(result);
}

export async function setPullRequestThreadStatus(
  input: SetPullRequestThreadStatusInput,
): Promise<PrThread> {
  const result = await invokeCommand("set_pull_request_thread_status", { input });
  return prThreadSchema.parse(result);
}

export async function submitPullRequestVote(
  input: SubmitPullRequestVoteInput,
): Promise<PrReviewer> {
  const result = await invokeCommand("submit_pull_request_vote", { input });
  return prReviewerSchema.parse(result);
}

export async function updatePullRequest(input: {
  organizationId?: string;
  projectId: string;
  repositoryId: string;
  pullRequestId: number;
  action: PullRequestAction;
  mergeStrategy?: string;
  deleteSourceBranch?: boolean;
  transitionWorkItems?: boolean;
}): Promise<PrStatusResult> {
  const result = await invokeCommand("update_pull_request", { input });
  return prStatusResultSchema.parse(result);
}

export async function setPullRequestReviewerRequired(input: {
  organizationId?: string;
  projectId: string;
  repositoryId: string;
  pullRequestId: number;
  reviewerId: string;
  isRequired: boolean;
}): Promise<void> {
  await invokeCommand("set_pull_request_reviewer_required", { input });
}

export async function removePullRequestReviewer(input: {
  organizationId?: string;
  projectId: string;
  repositoryId: string;
  pullRequestId: number;
  reviewerId: string;
}): Promise<void> {
  await invokeCommand("remove_pull_request_reviewer", { input });
}

export async function updatePullRequestDetails(input: {
  organizationId?: string;
  projectId: string;
  repositoryId: string;
  pullRequestId: number;
  title: string;
  description?: string;
}): Promise<PrDetailsResult> {
  const result = await invokeCommand("update_pull_request_details", { input });
  return prDetailsResultSchema.parse(result);
}

export async function searchPullRequestMentions(
  input: SearchPullRequestMentionsInput,
): Promise<MentionCandidate[]> {
  const result = await invokeCommand("search_pull_request_mentions", { input });
  return mentionCandidatesSchema.parse(result);
}

export async function editPullRequestComment(
  input: EditPullRequestCommentInput,
): Promise<PrThread> {
  const result = await invokeCommand("edit_pull_request_comment", { input });
  return prThreadSchema.parse(result);
}

export async function deletePullRequestComment(
  input: DeletePullRequestCommentInput,
): Promise<void> {
  await invokeCommand("delete_pull_request_comment", { input });
}

File: src-tauri/src/commands/pr_review.rs
use tauri::State;

use crate::app_state::{ensure_write_enabled, AppState};
use crate::error::Result;
use crate::pr_review::{
    DeletePullRequestCommentInput, EditPullRequestCommentInput, GetPullRequestFileDiffInput,
    PostPullRequestCommentInput, PrCommit, PrDetailsResult, PrFileDiff, PrLocator, PrReviewer,
    PrStatusResult, PrThread, PullRequestChanges, PullRequestReview,
    RemovePullRequestReviewerInput, SearchPullRequestMentionsInput,
    SetPullRequestReviewerRequiredInput, SetPullRequestThreadStatusInput,
    SubmitPullRequestVoteInput, UpdatePullRequestDetailsInput, UpdatePullRequestInput,
};
use crate::work_items::MentionCandidate;

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn get_pull_request_review(
    input: PrLocator,
    state: State<'_, AppState>,
) -> Result<PullRequestReview> {
    state.provider().await?.get_pull_request_review(input).await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn list_pull_request_changes(
    input: PrLocator,
    state: State<'_, AppState>,
) -> Result<PullRequestChanges> {
    state
        .provider()
        .await?
        .list_pull_request_changes(input)
        .await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn get_pull_request_file_diff(
    input: GetPullRequestFileDiffInput,
    state: State<'_, AppState>,
) -> Result<PrFileDiff> {
    state
        .provider()
        .await?
        .get_pull_request_file_diff(input)
        .await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn list_pull_request_commits(
    input: PrLocator,
    state: State<'_, AppState>,
) -> Result<Vec<PrCommit>> {
    state
        .provider()
        .await?
        .list_pull_request_commits(input)
        .await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn post_pull_request_comment(
    input: PostPullRequestCommentInput,
    state: State<'_, AppState>,
) -> Result<PrThread> {
    ensure_write_enabled(&state).await?;
    state
        .provider()
        .await?
        .post_pull_request_comment(input)
        .await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn set_pull_request_thread_status(
    input: SetPullRequestThreadStatusInput,
    state: State<'_, AppState>,
) -> Result<PrThread> {
    ensure_write_enabled(&state).await?;
    state
        .provider()
        .await?
        .set_pull_request_thread_status(input)
        .await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn submit_pull_request_vote(
    input: SubmitPullRequestVoteInput,
    state: State<'_, AppState>,
) -> Result<PrReviewer> {
    ensure_write_enabled(&state).await?;
    state
        .provider()
        .await?
        .submit_pull_request_vote(input)
        .await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn update_pull_request(
    input: UpdatePullRequestInput,
    state: State<'_, AppState>,
) -> Result<PrStatusResult> {
    ensure_write_enabled(&state).await?;
    state.provider().await?.update_pull_request(input).await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn set_pull_request_reviewer_required(
    input: SetPullRequestReviewerRequiredInput,
    state: State<'_, AppState>,
) -> Result<()> {
    ensure_write_enabled(&state).await?;
    state
        .provider()
        .await?
        .set_pull_request_reviewer_required(input)
        .await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn remove_pull_request_reviewer(
    input: RemovePullRequestReviewerInput,
    state: State<'_, AppState>,
) -> Result<()> {
    ensure_write_enabled(&state).await?;
    state
        .provider()
        .await?
        .remove_pull_request_reviewer(input)
        .await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn update_pull_request_details(
    input: UpdatePullRequestDetailsInput,
    state: State<'_, AppState>,
) -> Result<PrDetailsResult> {
    ensure_write_enabled(&state).await?;
    state
        .provider()
        .await?
        .update_pull_request_details(input)
        .await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn search_pull_request_mentions(
    input: SearchPullRequestMentionsInput,
    state: State<'_, AppState>,
) -> Result<Vec<MentionCandidate>> {
    state
        .provider()
        .await?
        .search_pull_request_mentions(input)
        .await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn edit_pull_request_comment(
    input: EditPullRequestCommentInput,
    state: State<'_, AppState>,
) -> Result<PrThread> {
    ensure_write_enabled(&state).await?;
    state
        .provider()
        .await?
        .edit_pull_request_comment(input)
        .await
}

#[tauri::command]
#[tracing::instrument(skip(state))]
pub async fn delete_pull_request_comment(
    input: DeletePullRequestCommentInput,
    state: State<'_, AppState>,
) -> Result<()> {
    ensure_write_enabled(&state).await?;
    state
        .provider()
        .await?
        .delete_pull_request_comment(input)
        .await
}
